README.rst
python-bluefin 1.4
==================

python-bluefin is a Python API client for the Bluefin_ Payment System API.
The only thing this client module does is serialize and pass the data on to
the Bluefin API gateway, it performs little to no validation. Values are passed
into an API function in dict form, and responses come back in a similar
dict format.

.. note:: python-bluefin is not at all endorsed by Bluefin Payment Systems in
    any way. We have published this API module with the hope that it will be
    useful to someone else.

.. _Bluefin: http://www.bluefin.com/

Status
------

python-bluefin has been in use in production for about a year with a high
transaction volume. The directmode client in particular has seen lots of use.

That said, we only use a little sliver of what Bluefin supports (one-off
credit card charges, namely). Given that python-bluefin is a very light
wrapper, other usage cases should still work just fine, even if we (the
maintainers) haven't used them yet.

Installing
----------

To install::

    pip install --upgrade bluefin

A quick example
---------------

Here's a very bare-bones example of how to use the Direct Mode V3Client. We
pass in the required key/values via a Python dict, and get a dict result. All
of these values are documented in the Bluefin API documentation, so look there
for help on just what is being passed in and coming back.

    >>> from bluefin.directmode.clients import V3Client
    >>> api = V3Client()
    >>> result = api.send_request({
        'pay_type': 'C',
        'tran_type': 'A',
        'account_id': 123456789012,
        'amount': 1.0,
        'card_number': 4444333322221111,
        'card_expire': '1212',
        'dynip_sec_code': 'SECURITY_CODE_HERE',
    })
    >>> print result
    {
        'avs_code': 'X', 'auth_msg': 'TEST APPROVED',
        'status_code': 'T', 'ticket_code': 'XXXXXXXXXXXXXXX',
        'auth_date': '2011-06-22 19:04:30', 'settle_currency': 'USD',
        'auth_code': '999999', 'settle_amount': '1', 'cvv2_code': 'M',
        'processor': 'TEST', 'trans_id': '123456789012'
    }


Running Unit Tests
------------------

* Install nose.
* Copy ``tests/api_details.blank.py`` to ``tests/api_details.py``
* Edit ``test/api_details.py`` to reflect your account number and security code.
* From within the ``python-bluefin`` dir, run ``nosetests``
  
License
-------

python-bluefin is licensed under the `BSD License`_.

.. _BSD License: https://github.com/duointeractive/python-bluefin/blob/master/LICENSE


CHANGES.txt
Changelog
=========

1.4
---

* Added an auto-retry mechanism to directmode. This is currently only used
  to retry when we encounter HTTP 408's from Bluefin.
* Added a max_retries keyword to V3Client's constructor. Defaults to 3 retries.

1.3
---

* Added a new V3ClientDeclinedException exception for directmode. This is
  raised when Bluefin sets an error code on an alternate 'status_code'
  attribute in the response.
* The 'message' attribute on the V3ClientDeclinedException features a
  "human-friendly" message. Bluefin spits out some pretty cryptic junk for
  the 'status_code' induced errors.
* To get the un-modified error message that came back from the Bluefin API
  (for any of our V3Client exceptions), you can now use the `raw_message`
  attribute on the exception. The `message` attribute is now replaced with a
  more friendly version for certain crappy error messages from Bluefin.

1.2
---

* Moved to the excellent requests library instead of urllib2.
* Error handling raises exceptions for directmode status_codes that are
  known to mark failed queries.

1.1
---

* Adding dataretrieval module.
* Additional unit tests.

1.0
---

* Initial release.

tests/directmode_tests.py
import unittest
from bluefin.directmode.clients import V3Client
from bluefin.directmode.exceptions import V3ClientInputException, V3ClientProcessingException, V3ClientDeclinedException
from tests.api_details import API_DETAILS, TEST_CARD, INVALID_CARD_NUM

class SaleTests(unittest.TestCase):
    """
    Tests for the 'sale' transaction type.
    """
    def setUp(self):
        self.api = V3Client()

    def test_basic(self):
        """
        Test a basic successful API call.
        """
        response = self.api.send_request({
            'pay_type': 'C',
            'tran_type': 'S',
            'account_id': API_DETAILS['account_id'],
            'amount': 1.0,
            'card_number': TEST_CARD['card_number'],
            'card_expire': TEST_CARD['card_expire'],
            'disable_cvv2': 'true',
            'dynip_sec_code': API_DETAILS['dynip_sec_code'],
            'bill_name1': 'First',
            'bill_name2': 'Last',
            'bill_street': '123 Some Street',
            'bill_zip': '12345',
            # Two-letter country code.
            # http://www.iso.org/iso/en/prods-services/iso3166ma/index.html
            'bill_country': 'US',
        })
        self.assertEqual(response['status_code'], '1', response)

    def test_invalid_cc_num(self):
        """
        Uses an invalid CC num to cause an error.
        """
        api_values = {
            'pay_type': 'C',
            'tran_type': 'S',
            'account_id': API_DETAILS['account_id'],
            'amount': 1.0,
            'card_number': INVALID_CARD_NUM,
            'card_expire': TEST_CARD['card_expire'],
            'disable_cvv2': 'true',
            'dynip_sec_code': API_DETAILS['dynip_sec_code'],
            'bill_name1': 'First',
            'bill_name2': 'Last',
            'bill_street': '123 Some Street',
            'bill_zip': '12345',
            # Two-letter country code.
            # http://www.iso.org/iso/en/prods-services/iso3166ma/index.html
            'bill_country': 'US',
        }
        self.assertRaises(V3ClientDeclinedException,
            self.api.send_request,
                api_values)


class AuthorizeTests(unittest.TestCase):
    """
    Tests for authorization API calls.
    """
    def setUp(self):
        self.api = V3Client()

    def test_basic(self):
        """
        Test a basic successful API call.
        """
        self.api.send_request({
            'pay_type': 'C',
            'tran_type': 'A',
            'account_id': API_DETAILS['account_id'],
            'amount': 1.0,
            'card_number': TEST_CARD['card_number'],
            'card_expire': TEST_CARD['card_expire'],
            'disable_cvv2': 'true',
            'dynip_sec_code': API_DETAILS['dynip_sec_code'],
            'bill_name1': 'First',
            'bill_name2': 'Last',
            'bill_street': '123 Some Street',
            'bill_zip': '12345',
            # Two-letter country code.
            # http://www.iso.org/iso/en/prods-services/iso3166ma/index.html
            'bill_country': 'US',
        })

    def test_invalid_sec_code(self):
        """
        Use an invalid security code to test exception handling.
        """
        api_values = {
            'pay_type': 'C',
            'tran_type': 'A',
            'account_id': API_DETAILS['account_id'],
            'amount': 1.0,
            'card_number': TEST_CARD['card_number'],
            'card_expire': TEST_CARD['card_expire'],
            'dynip_sec_code': 'HUZZAH_FOR_I_AM_INVALID',
        }

        self.assertRaises(V3ClientInputException, self.api.send_request, api_values)


    def test_missing_cc_num(self):
        api_values = {
            'pay_type': 'C',
            'tran_type': 'A',
            'account_id': API_DETAILS['account_id'],
            'amount': 1.0,
            #'card_number': TEST_CARD['card_number'],
            'card_expire': TEST_CARD['card_expire'],
            'dynip_sec_code': API_DETAILS['dynip_sec_code'],
            }

        self.assertRaises(V3ClientInputException, self.api.send_request,
                          api_values)

    def test_missing_cc_expiration(self):
        api_values = {
            'pay_type': 'C',
            'tran_type': 'A',
            'account_id': API_DETAILS['account_id'],
            'amount': 1.0,
            'card_number': TEST_CARD['card_number'],
            #'card_expire': TEST_CARD['card_expire'],
            'dynip_sec_code': API_DETAILS['dynip_sec_code'],
            }

        self.assertRaises(V3ClientInputException, self.api.send_request,
                          api_values)

    def test_missing_amount(self):
        api_values = {
            'pay_type': 'C',
            'tran_type': 'A',
            'account_id': API_DETAILS['account_id'],
            #'amount': 1.0,
            'card_number': TEST_CARD['card_number'],
            'card_expire': TEST_CARD['card_expire'],
            'dynip_sec_code': API_DETAILS['dynip_sec_code'],
            }

        self.assertRaises(V3ClientInputException, self.api.send_request,
                          api_values)

    def test_invalid_amount(self):
        api_values = {
            'pay_type': 'C',
            'tran_type': 'A',
            'account_id': API_DETAILS['account_id'],
            'amount': -1.0,
            'card_number': TEST_CARD['card_number'],
            'card_expire': TEST_CARD['card_expire'],
            'dynip_sec_code': API_DETAILS['dynip_sec_code'],
            }

        self.assertRaises(V3ClientInputException, self.api.send_request,
                          api_values)

        # Now set it to something non-numerical.
        api_values['amount'] = 'a'
        self.assertRaises(V3ClientInputException, self.api.send_request,
                          api_values)

class AuthorizeRecurringTests(unittest.TestCase):
    """
    Unit tests for recurring payment authorization.
    """
    def setUp(self):
        self.api = V3Client()

    def test_basic(self):
        """
        Test a basic successful API call.
        """
        self.api.send_request({
            'pay_type': 'C',
            'tran_type': 'A',
            'account_id': API_DETAILS['account_id'],
            'amount': 1.0,
            'recurring_amount': 1.0,
            'recurring_period': 31,
            'card_number': TEST_CARD['card_number'],
            'card_expire': TEST_CARD['card_expire'],
            'dynip_sec_code': API_DETAILS['dynip_sec_code'],
            })

tests/api_details.blank.py
"""
The following credentials are used to run the unit tests. Please copy this
file to api_details.py and modify the values to fit your account.
"""
API_DETAILS = {
    # You can get this from your account dashboard on secure.bluefingateway.com.
    'account_id': 123456789012,
    # Under Setup -> Account Security.
    'dynip_sec_code': 'YOUR_CODE_HERE',
    # For dataretrieval tests, navigate to your Site Tags setup and pick the
    # auth code out.
    'site_tag_auth_code': 'YOUR_AUTH_CODE_HERE'
}

# Make sure you enter your test credit card here. Under Setup -> Credit Cards.
TEST_CARD = {
    'card_number': 4444333322221111,
    'card_expire': '1212',
}

# An invalid card number that doesn't match your test card number.
INVALID_CARD_NUM = 4012888888881881
# This has to not equal your test card number in order for it to fail.
assert INVALID_CARD_NUM != TEST_CARD['card_number']

tests/dataretrival_tests.py
import unittest
from bluefin.dataretrieval.clients import V1Client
from bluefin.dataretrieval.exceptions import V1ClientInputException, V1ClientProcessingException
from tests.api_details import API_DETAILS, TEST_CARD

class TransactionReportingTest(unittest.TestCase):
    """
    Tests for transaction reporting API calls.
    """
    def test_basic(self):
        """
        Test a basic successful API call.
        """
        api = V1Client()
        result = api.send_request({
            'transactions_after': '2006-12-30',
            'account_id': API_DETAILS['account_id'],
            'site_tag': 'MAIN',
            'authorization': API_DETAILS['site_tag_auth_code'],
        })

    def test_no_input(self):

        api = V1Client()
        self.assertRaises(V1ClientInputException, api.send_request, {})




bluefin/directmode/exceptions.py
"""
Exception classes for Direct Mode API operations.
"""
class V3ClientException(Exception):
    """
    Used to denote some kind of generic error. This does not include errors
    returned from Bluefin API responses. Those are handled by the more
    specific exception classes below.
    """

    def __init__(self, message, error_code=None, *args):
        Exception.__init__(self, message, error_code, *args)
        # The error message back from the Bluefin API server.
        self.message = self.clean_error_message(message)
        # And if you like it raw...
        self.raw_message = message
        # The HTTP error code returned.
        self.error_code = error_code

    def __str__(self):
        if self.error_code:
            return "%s (Error Code: %s)" % (self.message, self.error_code)
        else:
            return self.message

    def clean_error_message(self, message):
        """
        Some of Bluefin's errors are pretty bad, particularly the reason_code2
        errors. We provide a stub here to make some of the really awful
        error messages more bearable. This is currently only used
        in V3ClientDeclinedException, but may grow with time as more
        hideous errors are encountered.

        :rtype: basestring
        :returns: A 'prettied up' error message.
        """

        return message


class V3ClientInputException(V3ClientException):
    """
    Raised when HTTP error codes 600-699 are returned, indicating an input
    issue of some sort.
    """

    pass


class V3ClientProcessingException(V3ClientException):
    """
    Raised when a processing error occurs on the Bluefin side. These are
    generally HTTP error codes 700-799.
    """

    pass


class V3ClientDeclinedException(V3ClientProcessingException):
    """
    Bluefin has a wonky additional 'status_code' return value that is used
    separately from HTTP status codes. These look to all be authorization
    related so far.
    """

    def clean_error_message(self, message):
        """
        These 'status_code' errors have hideous error messages that just
        aren't good to show users. We'll pretty them up here. Whatever is
        returned here becomes :py:attr:`message`, while the original gets
        set to :py:attr:`raw_message`.

        :rtype: basestring
        """

        if not message:
            # Can't do much here.
            return message

        if message.startswith("CVV2"):
            return "The Card Security Code that was provided is invalid. " \
                   "Please check the three-digit security code on the back of " \
                   "your card and try again."
        elif message.startswith("INVALID CARD NO"):
            return "The credit card number that was provided is invalid. Please "\
                   "re-enter and try again."
        elif message.startswith("CVD"):
            return "Invalid card number, security code, or other value. Please " \
                   "check your input and try again."
        elif message.startswith("C/DECLINED"):
            return "Your payment was declined."
        elif message.startswith("AUTH DECLINED"):
            return "Your payment was declined."

        return message

bluefin/directmode/clients.py
"""
Client classes for Direct Mode services.
"""

import urlparse
import requests

from bluefin.directmode.exceptions import V3ClientInputException, V3ClientProcessingException, V3ClientException, V3ClientDeclinedException

class V3Client(object):
    """
    This is the class used to send API calls and receive responses through for
    the V3.x Direct Mode API client.
    """
    def __init__(self, host='https://secure.bluefingateway.com:1402',
                 path='/gw/sas/direct3.1', http_timeout=15,
                 account_id=None, dynip_sec_code=None, max_retries=3):
        """
        Instantiates our API interface with sensible defaults.

        :keyword str host: Full URI to the API gateway, including protocol,
            hostname, and port. No trailing slash.
        :keyword str path: The path to the API endpoint.
        :keyword int http_timeout: Socket timeout in seconds. This is globally
            applied, so be careful.
        :keyword int account_id: Your Bluefin account number. Providing this
            value here means you don't have to pass it with each value dict
            to :py:meth:`send_request`.
        :keyword str dynip_sec_code: A dynamic IP security code. Providing this
            value here means you don't have to pass it with each value dict
            to :py:meth:`send_request`.
        :keyword int max_retries: Maximum number of retries in the event we
            run into a retryable error.
        """

        # Default to the HTTPS endpoint.
        self.host = host
        # Default to the Transaction interface.
        self.path = path
        # Note that this is applied globally, so be careful.
        self.http_timeout = http_timeout
        # Maximum number of retries in the event we run into a retryable error.
        self.max_retries = max_retries

        self.default_values = {}

        if account_id:
            self.default_values['account_id'] = account_id
        if dynip_sec_code:
            self.default_values['dynip_sec_code'] = dynip_sec_code

    def _get_endpoint(self):
        """
        urllib2.Request wants a full URI with protocol, host, and path.
        Assemble the proto+host+path into a URI to request.
        """
        return '%s%s' % (self.host, self.path)

    def _check_for_error_http_status_code(self, response):
        """
        Checks the response's HTTP status code for common error code numbers.

        :param requests.Response response: A requests Response object generated
            by requests.post().
        :raises: An appropriate bluefin.directmode.exceptions.V3ClientException
            sub-class, depending on the error.
        """
        http_status = response.status_code

        if http_status >= 600 and http_status <= 699:
            # HTTP error codes 600-699 are input errors.
            raise V3ClientInputException(response.text, error_code=http_status)
        elif http_status >= 700 and http_status <= 799:
            # HTTP error codes 700-799 http_status processing errors.
            raise V3ClientProcessingException(response.text, error_code=http_status)
        elif http_status > 200:
            raise V3ClientException(response.text, error_code=http_status)

    def _check_parsed_response_for_error_codes(self, result_dict):
        """
        Looks through the dict that we get by parsing the response from Bluefin.
        Finds any known error conditions, raises an appropriate exception.

        :param dict result_dict: The urlparsed response from Bluefin.
        :raises: An appropriate bluefin.directmode.exceptions.V3ClientException
            sub-class, depending on the error.
        """

        status_code = result_dict.get('status_code')
        # These are the two status codes that indicate issues.
        if result_dict.get('status_code') in ['F', '0']:
            reason_code = result_dict.get('reason_code2')
            # There are multiple error fields to check. auth_msg is almost always
            # the one to go by.
            err_message = result_dict.get('auth_msg') or reason_code
            # These errors don't tend to use HTTP status codes, so check the
            # reason_code2 field.
            exc_code = status_code or reason_code
            raise V3ClientDeclinedException(err_message, error_code=exc_code)

    def send_request(self, values):
        """
        Sends an API request. You are on your own to pass in the correct
        key/value pairs as a dict in the ``values`` argument.

        :param dict values: Key/value pairs for your desired API call. See
            the Bluefin documentation for what these should be.
        :rtype: dict
        :returns: A dict of output from the API server. See the Bluefin API
            docs for how to interpret this.
        :raises: V3ClientInputException when the Bluefin API says we have
            an input error, and V3ClientProcessingException when the Bluefin
            API encounters an error during processing. The lower level
            urllib2 may raise urllib2.HTTPError exceptions also.
        """

        # Copy the default values dict so we don't have to repeat stuff like
        # account_id and dynip_sec_code for every request.
        all_values = self.default_values.copy()
        # The transaction values can override the defaults.
        all_values.update(values)

        retries = 0
        response = None
        # I hate to retry within an infinite loop, but it avoids recursion,
        # and it works.
        while True:
            response = requests.post(
                self._get_endpoint(),
                data=all_values,
                timeout=self.http_timeout
            )

            # Looks at the HTTP status code and raises an exception if any of the
            # known number ranges for errors are returned.
            try:
                self._check_for_error_http_status_code(response)
            except V3ClientException, exc:
                if retries >= self.max_retries:
                    # We've exceeded the max number of retries. We don't need
                    # to see what error this is, because we can't retry any more.
                    raise

                if exc.error_code in [408, '408']:
                    # 408's are often network related, and are safe to retry.
                    # Increment the retry counter and jump back to the top
                    # of the loop.
                    retries += 1
                    continue
                else:
                    # This is not a retryable error. Re-raise that sucker.
                    raise

            # Nothing bad happened. Break the loop.
            break

        result_dict = urlparse.parse_qs(response.text)
        for key, value in result_dict.items():
            # Strip away the lists from the value, since these should all just
            # be a one-member list. We'll join with commas just in case.
            result_dict[key] = ','.join(value)

        # Looks through the parsed response dict for common error codes. Raises
        # exceptions if any are found.
        self._check_parsed_response_for_error_codes(result_dict)

        return result_dict